lib/elixir/lib/exception.ex

```diff
@@ -237,6 +241,67 @@ defmodule Exception do
     end
   end
 
+  defp is_map_node?({:is_map, _, [_]}), do: true
+  defp is_map_node?(_), do: false
+  defp is_map_key_node?({:is_map_key, _, [_, _]}), do: true
+  defp is_map_key_node?(_), do: false
+
+  defp struct_validation_node?(
+         {:is_atom, _, [{{:., [], [:erlang, :map_get]}, _, [:__struct__, _]}]}
+       ),
+       do: true
+
+  defp struct_validation_node?(
+         {:==, _, [{{:., [], [:erlang, :map_get]}, _, [:__struct__, _]}, _module]}
+       ),
+       do: true
+
+  defp struct_validation_node?(_), do: false
+
+  defp is_struct_macro?(
+         {:and, _, [{:and, _, [%{node: node_1}, %{node: node_2}]}, %{node: node_3}]}
+       ),
+       do: is_map_node?(node_1) and is_map_key_node?(node_2) and struct_validation_node?(node_3)
+
+  defp is_struct_macro?(
+         {:and, _,
+          [
+            {:and, _,
+             [
+               {:and, _,
+                [%{node: node_1}, {:or, _, [%{node: {:is_atom, _, [_]}}, %{node: :fail}]}]},
+               %{node: node_2}
+             ]},
+            %{node: node_3}
+          ]}
+       ),
+       do: is_map_node?(node_1) and is_map_key_node?(node_2) and struct_validation_node?(node_3)
+
+  defp is_struct_macro?(_), do: false
+
+  defp translate_guard(guard) do
+    cond do
+      is_struct_macro?(guard) ->
+        undo_guard_macro(:is_struct, guard)
+
+      true ->
+        guard
+    end
+  end
+
+  defp undo_guard_macro(
+         :is_struct,
+         {:and, meta, [_, %{node: {_, _, [{_, _, [_, {struct, _, _}]} | optional]}}]}
+       ) do
```

Comment:
```suggestion
  defp translate_guard(guard) do
    if is_struct_macro?(guard) do
      undo_is_struct_guard(guard)
    else
      guard
    end
  end

  defp undo_is_struct_guard(
         {:and, meta, [_, %{node: {_, _, [{_, _, [_, {struct, _, _}]} | optional]}}]}
       ) do
```

Reply:
I added the cond there in case there are more places other than is_struct that use macros for guards, then I thought that it would be a good place to add the check for them. Same for the undo_guard_macro, by pattern matching with the guard name. But it was just a guess that there would be more cases, I haven't found any 😂